File: haku-app/src/etfs/etfPrototyping/getCagr.js
// Get Cagr

import mainData from "./sample-monthly-values.json" assert { type: "json" };

for (let etf in mainData) {
  const monthlyTimeSeries = mainData[etf]["Monthly Adjusted Time Series"];

  let monthlyKeys = Object.keys(monthlyTimeSeries);

  let currentDate = new Date(monthlyKeys[0]);

  let fundStartDate = new Date(monthlyKeys[monthlyKeys.length - 1]);

  let fundDuration =
    (currentDate.getTime() - fundStartDate.getTime()) /
    (1000 * 60 * 60 * 24 * 365);

  let diff =
    monthlyTimeSeries[monthlyKeys[0]]["4. close"] /
    monthlyTimeSeries[monthlyKeys[monthlyKeys.length - 1]]["4. close"];

  let etfCagr = diff ** (1 / fundDuration) - 1;

  mainData[etf]["Meta Data"]["5. CAGR"] = etfCagr;

  console.log(mainData[etf]["Meta Data"]);
}

/*
const monthlyValues = mainData["ARKK"]["Monthly Adjusted Time Series"];

let monthKeys = Object.keys(monthlyValues);

let startDate = new Date(monthKeys[0]);
let endDate = new Date(monthKeys[monthKeys.length - 1]);

let years =
  (startDate.getTime() - endDate.getTime()) / (1000 * 60 * 60 * 24 * 365);

console.log("Fund Years", years);

const diffCuota =
  monthlyValues[monthKeys[0]]["4. close"] /
  monthlyValues[monthKeys[monthKeys.length - 1]]["4. close"];

console.log("Fund Diff: ", diffCuota);

let cagr = diffCuota ** (1 / years) - 1;
console.log("Cagr: ", cagr);

*/

File: haku-data/ETFs/Prototypes2/get-cagr.py
import pandas as pd
import json
from datetime import datetime
import os
print("Current working directory:", os.getcwd())   



with open("./haku-data/ETFs/Prototypes2/ten-etf-monthly-values.json", "r") as file:
    data = json.load(file) 


for etf in data:
    monthly_values = data[etf]["Monthly Adjusted Time Series"]
    
    monthly_values_keys = list(monthly_values.keys())

    current_date = datetime.strptime(monthly_values_keys[0],  "%Y-%m-%d").date()
    start_date = datetime.strptime(monthly_values_keys[-1],  "%Y-%m-%d").date()
    
    duration = (current_date - start_date) # duration is a datetime object 

    fund_years = duration.days / 365.25
    
    current_etf_price = float(monthly_values[monthly_values_keys[0]]["4. close"])

    start_etf_price = float(monthly_values[monthly_values_keys[-1]]["4. close"])

    cagr = ((current_etf_price / start_etf_price)**(1/fund_years)) - 1

    data[etf]["Meta Data"]["5. CAGR"] = cagr

    print(f"{etf} CAGR: ", cagr)

print(data['ARKK']["Meta Data"])

# Store data in JSON file
with open('ten-etf-monthly-values.json', 'w') as json_file:
    json.dump(data, json_file, indent=4)

